credentials.go

```diff
@@ -0,0 +1,107 @@
+// This source file is part of the EdgeDB open source project.
+//
+// Copyright 2020-present EdgeDB Inc. and the EdgeDB authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package edgedb
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"math"
+)
+
+type credentials struct {
+	host     string
+	port     int
+	user     string
+	database string
+	password string
+}
+
+func readCredentials(path string) (*credentials, error) {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"cannot read credentials at %q: %v%w", path, err, ErrBadConfig,
+		)
+	}
+
+	var values map[string]interface{}
+	if e := json.Unmarshal(data, &values); e != nil {
+		return nil, fmt.Errorf(
+			"cannot read credentials at %q: %v%w", path, e, ErrBadConfig,
+		)
+	}
+
+	creds, err := validateCredentials(values)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"cannot read credentials at %q: %v%w", path, err, ErrBadConfig,
+		)
+	}
+
+	return creds, nil
+}
+
+func validateCredentials(data map[string]interface{}) (*credentials, error) {
```

Comment:
I see an issue with defaults here. E.g. default user and database are `edgedb`, default port is 5656.

The problem with applying these defaults later is that, config like this:
```
{ database: "" }
```
Would be treated differently in no an non go. In all other languages it's the error (database name is empty). But because golang doesn't distinguish between empty string and non-existing string, it will be equivalent to `database: "edgedb"`.